drivers/wifi/esp/esp.h

```diff
@@ -218,35 +229,110 @@ struct esp_data {
 
 int esp_offload_init(struct net_if *iface);
 
-struct esp_socket *esp_socket_get();
+struct esp_socket *esp_socket_get(struct esp_data *data,
+				  enum net_sock_type type,
+				  enum net_ip_protocol ip_proto,
+				  struct net_context *context);
 int esp_socket_put(struct esp_socket *sock);
 struct esp_socket *esp_socket_from_link_id(struct esp_data *data,
 					   uint8_t link_id);
 void esp_socket_init(struct esp_data *data);
 void esp_socket_close(struct esp_socket *sock);
 void esp_socket_rx(struct esp_socket *sock, struct net_buf *buf,
 		   size_t offset, size_t len);
+void esp_socket_workq_stop_and_flush(struct esp_socket *sock);
+struct esp_socket *esp_socket_ref(struct esp_socket *sock);
+void esp_socket_unref(struct esp_socket *sock);
 
-void esp_socket_workq_flush(struct esp_socket *sock);
+static inline
+struct esp_socket *esp_socket_ref_from_link_id(struct esp_data *data,
+					       uint8_t link_id)
+{
+	return esp_socket_ref(esp_socket_from_link_id(data, link_id));
```

Comment:
Need to check for `NULL` before calling `esp_socket_ref` here?

Reply:
Good catch. `esp_socket_from_link_id()` was always returning non-NULL pointer, so that was not causing a problem. The problem was that this pointer was not always valid, as `link_id` was not verified. So I decided to drop separate `esp_socket_from_link_id()` function and simply do:
```
static inline
struct esp_socket *esp_socket_ref_from_link_id(struct esp_data *data,
					       uint8_t link_id)
{
	if (link_id >= ARRAY_SIZE(data->sockets)) {
		return NULL;
	}

	return esp_socket_ref(&data->sockets[link_id]);
}
```